lib/active_model/serializer.rb

```diff
@@ -112,10 +112,24 @@ def self.digest_caller_file(caller_line)
       Digest::MD5.hexdigest(serializer_file_contents)
     end
 
+    def self.serializer_lookup_chain_for(klass)
+      chain = []
+
+      resource_class_name = klass.name.demodulize
+      resource_namespace = klass.name.deconstantize
+      serializer_class_name = "#{resource_class_name}Serializer"
+
+      chain.push("#{name}::#{serializer_class_name}") if self != ActiveModel::Serializer
+      chain.push("#{resource_namespace}::#{serializer_class_name}")
+      chain.push(serializer_class_name)
+
+      chain
+    end
+
     def self.get_serializer_for(klass)
       serializers_cache.fetch_or_store(klass) do
-        serializer_class_name = "#{klass.name}Serializer"
-        serializer_class = serializer_class_name.safe_constantize
+        # NOTE(beauby): When we drop 1.9.3 support we can lazify the map for perfs.
+        serializer_class = serializer_lookup_chain_for(klass).map(&:safe_constantize).find { |x| x }
```

Comment:
For naming, `serializer_lookup_chain` refers to the "lookup chain" for serializers, i.e. the chain of possibilities used for looking up the right serializer (and not a "lookups chain" which would be a chain of lookups, which I'm not even sure what it would mean).

As for functional vs ad-hoc, I'm referring to
```ruby
serializer_class = serializer_lookup_chain_for(klass).map(&:safe_constantize).find { |x| x }
```
vs
```ruby
        serializer_class = nil
        potential_serializer_class_names.each do |serializer_class_name|
          serializer_class = serializer_class_name.safe_constantize
          break if serializer_class
        end
```